pallets/msa/src/offchain_storage.rs

```diff
@@ -0,0 +1,135 @@
+/// Offchain Storage for MSA
+use codec::{Decode, Encode};
+use common_primitives::offchain as offchain_common;
+use sp_runtime::offchain::storage::StorageRetrievalError;
+use sp_std::{
+	collections::btree_map::BTreeMap,
+	fmt::{Debug, Formatter},
+	vec::Vec,
+};
+
+/// Generic prefix for MSA index storage
+pub const MSA_INDEX_KEY: &[u8] = b"frequency::msa::";
+
+/// Derive storage key for MSA index
+#[deny(clippy::clone_double_ref)]
+pub(crate) fn derive_storage_key<K>(msa_id: K) -> Vec<u8>
+where
+	K: Encode + Clone + Ord + Decode,
+{
+	[MSA_INDEX_KEY, msa_id.encode().as_slice()].concat().encode().to_vec()
+}
+
+/// MSA Public Key Data
+/// BTreeMap<MSA ID, Vec<Public Key>>  type structure
+/// Support scale encoding and decoding for efficient storage
+#[derive(Encode, Decode, Clone, PartialEq, Eq, Default)]
+pub struct MSAPublicKeyData<K, V>(pub BTreeMap<K, Vec<V>>);
+
```

Comment:
Currently the offchain storage stores an encoded ```MSAPublicKeyData``` which is just BTree of msa_id vs keys